src/ldapclient_lib.py

```diff
@@ -0,0 +1,395 @@
+
+
+class GlauthSnapReadyEvent(EventBase):
+    """Charm Event triggered when a glauth snap is ready to start."""
+
+    def __init__(
+        self,
+        handle: Handle,
+    ):
+        super().__init__(handle)
+
+
+class CertificateAvailableEvent(EventBase):
+    """Charm Event triggered when a CA certificate is available."""
+
+    def __init__(
+        self,
+        handle: Handle,
+        ca_cert: str,
+    ):
+        super().__init__(handle)
+        self.ca_cert = ca_cert
+
+    def snapshot(self) -> dict:
+        """Return snapshot."""
+        return {
+            "ca_cert": self.ca_cert,
+        }
+
+    def restore(self, snapshot: dict):
+        """Restore snapshot."""
+        self.ca_cert = snapshot["ca_cert"]
+
+
+class CertificateUnavailableEvent(EventBase):
+    """Charm Event triggered when a CA certificate is unavailable."""
+
+    pass
+
+
+class ConfigDataAvailableEvent(EventBase):
+    """Charm Event triggered when config data is available."""
+
+    def __init__(
+        self,
+        handle: Handle,
+        basedn: str,
+        ldap_uri: str,
+        ldbd_content: str,
+        lp_content: str,
+    ):
+        super().__init__(handle)
+        self.basedn = basedn
+        self.ldap_uri = ldap_uri
+        self.ldbd_content = ldbd_content
+        self.lp_content = lp_content
+
+    def snapshot(self) -> dict:
+        """Return snapshot."""
+        return {
+            "basedn": self.basedn,
+            "ldap_uri": self.ldap_uri,
+            "ldbd_content": self.ldbd_content,
+            "lp_content": self.lp_content,
+        }
+
+    def restore(self, snapshot: dict):
+        """Restore snapshot."""
+        self.basedn = snapshot["basedn"]
+        self.ldap_uri = snapshot["ldap_uri"]
+        self.ldbd_content = snapshot["ldbd_content"]
+        self.lp_content = snapshot["lp_content"]
+
+
+class ConfigDataUnavailableEvent(EventBase):
+    """Charm Event triggered when config data is unavailable."""
+
+    def __init__(
+        self,
+        handle: Handle,
+        api_port: int,
+        ldap_port: int,
+    ):
+        super().__init__(handle)
+        self.api_port = api_port
+        self.ldap_port = ldap_port
+
+    def snapshot(self) -> dict:
+        """Return snapshot."""
+        return {
+            "api_port": self.api_port,
+            "ldap_port": self.ldap_port,
+        }
+
+    def restore(self, snapshot: dict):
+        """Restore snapshot."""
+        self.api_port = snapshot["api_port"]
+        self.ldap_port = snapshot["ldap_port"]
+
+
+class ServerUnavailableEvent(EventBase):
+    """Charm Event triggered When the LDAP server is unavailable."""
+
+    def __init__(
+        self,
+        handle: Handle,
+    ):
+        super().__init__(handle)
+
+
+class LdapReadyEvent(EventBase):
+    """Charm Event triggered when LDAP is ready to start."""
+
+    def __init__(
+        self,
+        handle: Handle,
+    ):
+        super().__init__(handle)
```

Comment:
Same as my comment above:

```suggestion
class LdapReadyEvent(EventBase):
    """Charm Event triggered when LDAP is ready to start."""
```